Support frozen strings with REXML

frederikspang opened the pull request:
Tests are passing with `export RUBYOPT=--enable-frozen-string-literal`

johnnyshields commented on lib/onelogin/ruby-saml/utils.rb:
```diff
@@ -32,7 +32,7 @@ class Utils
         )
       $)x.freeze
-      UUID_PREFIX = '_'
+      UUID_PREFIX = +'_'
```
Constants should remain frozen. Instead, call `UUID_PREFIX.dup` where it is used, if necessaty.

frederikspang replied:
It's used with `.replace` which won't work. Reassigning the constant is "bad".
I'm replacing it with a variable, but should we keep it for compatibility, or remove it and bump a major version?

frederikspang commented:
@johnnyshields I've adjusted some stuff from review. UUID_PREFIX is used as a constant, and `.replace` with Utils.set_prefix, which is not working out great :). I've replaced it with a variable. Should we deprecate or remove the (now unused) constant?

johnnyshields commented on lib/onelogin/ruby-saml/utils.rb:
```diff
@@ -32,7 +32,9 @@ class Utils
       $)x.freeze
+
       UUID_PREFIX = '_'
+      @@prefix = '_'.freeze
```
not necessary to call .freeze here.

frederikspang replied:
@johnnyshields Fixed

johnnyshields commented:
@frederikspang yikes wow we were calling replace on a constant 🤦‍♂️ . The way you have it now with `@@prefix` is fine.

johnnyshields commented:
@pitbulk frozen string stuff is handled already on 2.x. This is a "backport" in case we want it on 1.x branch, but the alternative here would be to just close this PR and tell users that you must use 2.x for Ruby 3.4+. Which is reasonable in my opinion.

johnnyshields commented:
Mutable constant addressed here: https://github.com/SAML-Toolkits/ruby-saml/pull/735

frederikspang commented:
@johnnyshields Frozen strings is not exclusive for Ruby 3.4 - It's just moving towards default frozen, with "chilled string". 

I think that not supporting frozen strings for < 3.4 would be a mistake, when the backport is somewhat simple for 1.x branch.

frederikspang commented:
Specs may have been solved for older ruby versions, with using .dup instead of `+""` for unfreezing. Could you rerun specs?

johnnyshields commented:
@frederikspang we are close to releasing v2.x branch so this will be a moot point...

johnnyshields commented:
It's @pitbulk 's call whether he wants to release for 1.x branch.

pitbulk commented:
@frederikspang, it seems for ruby >= 3.4.0 we need to add mutex_m gem on the gemspec, can you add it?

frederikspang commented:
@pitbulk Got it :)

gstokkink commented:
@pitbulk can we get a new release for this? 😄
